File: evaluation_layer/eval_methods/Solution wizard evaluation/use cases/UC01_audio_incident_evaluation/scripts/build_workbook.py
#!/usr/bin/env python3
r"""Populate the UC01 Excel comparison template without external dependencies.

Place this file in the evaluation package's ``scripts`` directory and run it
from the package root:

    python .\scripts\build_workbook.py

Optional positional arguments:

    python .\scripts\build_workbook.py INPUT_JSON OUTPUT_XLSX
"""

from __future__ import annotations

import argparse
import json
import os
import re
import tempfile
import zipfile
from pathlib import Path
from typing import Any
from xml.etree import ElementTree as ET


MAIN_NS = "http://schemas.openxmlformats.org/spreadsheetml/2006/main"
DOC_REL_NS = "http://schemas.openxmlformats.org/officeDocument/2006/relationships"
PKG_REL_NS = "http://schemas.openxmlformats.org/package/2006/relationships"
NS = {"x": MAIN_NS, "r": DOC_REL_NS, "pr": PKG_REL_NS}
ET.register_namespace("x", MAIN_NS)
ET.register_namespace("r", DOC_REL_NS)

SHEET_GROUPS = {
    "EQ1 Requirements": "EQ1",
    "EQ1 Diagnostics": "EQ1_diagnostics",
    "EQ2 Configuration": "EQ2",
    "EQ3 Package": "EQ3",
    "EQ4 Execution": "EQ4",
}


def _excel_text(value: Any) -> str:
    """Return XML-safe text within Excel's 32,767-character cell limit."""

    if value is None:
        return ""
    if isinstance(value, (dict, list)):
        text = json.dumps(value, ensure_ascii=False)
    elif value is True:
        text = "Yes"
    elif value is False:
        text = "No"
    else:
        text = str(value)
    text = "".join(
        character
        for character in text
        if character in "\t\n\r" or ord(character) >= 32
    )
    if len(text) > 32767:
        text = f"{text[:32720]}\n[TRUNCATED FOR EXCEL CELL LIMIT]"
    return text


def _display_value(value: Any) -> str:
    if value is True:
        return "Yes"
    if value is False:
        return "No"
    if value is None:
        return ""
    return str(value)


def _sheet_paths(entries: dict[str, bytes]) -> dict[str, str]:
    workbook = ET.fromstring(entries["xl/workbook.xml"])
    relationships = ET.fromstring(entries["xl/_rels/workbook.xml.rels"])
    targets = {
        relationship.attrib["Id"]: relationship.attrib["Target"].lstrip("/")
        for relationship in relationships.findall("pr:Relationship", NS)
    }
    result: dict[str, str] = {}
    for sheet in workbook.findall("x:sheets/x:sheet", NS):
        relationship_id = sheet.attrib[f"{{{DOC_REL_NS}}}id"]
        target = targets[relationship_id]
        if not target.startswith("xl/"):
            target = f"xl/{target}"
        result[sheet.attrib["name"]] = target
    return result


def _cell(sheet: ET.Element, coordinate: str) -> ET.Element:
    found = sheet.find(f".//x:c[@r='{coordinate}']", NS)
    if found is not None:
        return found

    match = re.fullmatch(r"([A-Z]+)([0-9]+)", coordinate)
    if not match:
        raise ValueError(f"Invalid Excel coordinate: {coordinate}")
    row_number = match.group(2)
    sheet_data = sheet.find("x:sheetData", NS)
    if sheet_data is None:
        sheet_data = ET.SubElement(sheet, f"{{{MAIN_NS}}}sheetData")
    row = sheet_data.find(f"x:row[@r='{row_number}']", NS)
    if row is None:
        row = ET.SubElement(sheet_data, f"{{{MAIN_NS}}}row", {"r": row_number})
    return ET.SubElement(row, f"{{{MAIN_NS}}}c", {"r": coordinate})


def _set_text(sheet: ET.Element, coordinate: str, value: Any) -> None:
    cell = _cell(sheet, coordinate)
    style = cell.attrib.get("s")
    cell.clear()
    cell.attrib["r"] = coordinate
    if style is not None:
        cell.attrib["s"] = style
    text = _excel_text(value)
    cell.attrib["t"] = "str"
    if text:
        ET.SubElement(cell, f"{{{MAIN_NS}}}v").text = text


def _read_text(cell: ET.Element | None, shared_strings: list[str]) -> str:
    if cell is None:
        return ""
    cell_type = cell.attrib.get("t")
    if cell_type == "inlineStr":
        return "".join(
            node.text or "" for node in cell.findall(".//x:is//x:t", NS)
        )
    value = cell.find("x:v", NS)
    if value is None or value.text is None:
        return ""
    if cell_type == "s":
        try:
            return shared_strings[int(value.text)]
        except (ValueError, IndexError):
            return ""
    return value.text


def _shared_strings(entries: dict[str, bytes]) -> list[str]:
    data = entries.get("xl/sharedStrings.xml")
    if not data:
        return []
    root = ET.fromstring(data)
    return [
        "".join(node.text or "" for node in item.findall(".//x:t", NS))
        for item in root.findall("x:si", NS)
    ]


def _row_for_check(
    sheet: ET.Element, check_id: str, shared_strings: list[str]
) -> int:
    for cell in sheet.findall(".//x:c", NS):
        coordinate = cell.attrib.get("r", "")
        if not coordinate.startswith("B"):
            continue
        if _read_text(cell, shared_strings) == check_id:
            return int(coordinate[1:])
    raise KeyError(f"Check {check_id!r} is not present in the workbook template.")


def _run_results(comparison: dict[str, Any], run_name: str, group: str) -> dict[str, dict[str, Any]]:
    rows = (
        comparison.get("runs", {})
        .get(run_name, {})
        .get("results", {})
        .get(group, [])
    )
    return {
        str(row.get("check_id")): row
        for row in rows
        if isinstance(row, dict) and row.get("check_id")
    }


def _populate_evaluation_sheet(
    sheet: ET.Element,
    group: str,
    comparison: dict[str, Any],
    shared_strings: list[str],
) -> None:
    run_01 = _run_results(comparison, "run_01", group)
    run_02 = _run_results(comparison, "run_02", group)
    check_ids = sorted(set(run_01) | set(run_02))
    for check_id in check_ids:
        row_number = _row_for_check(sheet, check_id, shared_strings)
        first = run_01.get(check_id, {})
        second = run_02.get(check_id, {})
        values = {
            f"F{row_number}": first.get("generated_value", ""),
            f"G{row_number}": first.get("evidence", ""),
            f"H{row_number}": first.get("automatic_result", ""),
            f"I{row_number}": first.get("human_verdict", ""),
            f"J{row_number}": second.get("generated_value", ""),
            f"K{row_number}": second.get("evidence", ""),
            f"L{row_number}": second.get("automatic_result", ""),
            f"M{row_number}": second.get("human_verdict", ""),
            f"N{row_number}": " | ".join(
                str(note)
                for note in (
                    first.get("evaluator_notes", ""),
                    second.get("evaluator_notes", ""),
                )
                if note
            ),
        }
        for coordinate, value in values.items():
            _set_text(sheet, coordinate, value)


def _populate_recovery(sheet: ET.Element, comparison: dict[str, Any]) -> None:
    run_01 = (
        comparison.get("runs", {}).get("run_01", {}).get("recovery_diagnostic", {})
    )
    run_02 = (
        comparison.get("runs", {}).get("run_02", {}).get("recovery_diagnostic", {})
    )
    rows = (
        (4, "initial_execution_successful"),
        (5, "refinement_attempts_to_success"),
        (6, "final_execution_successful"),
        (7, "status"),
    )
    for row_number, key in rows:
        _set_text(sheet, f"B{row_number}", _display_value(run_01.get(key)))
        _set_text(sheet, f"C{row_number}", run_01.get("evidence", ""))
        _set_text(sheet, f"D{row_number}", _display_value(run_02.get(key)))
        _set_text(sheet, f"E{row_number}", run_02.get("evidence", ""))


def _force_recalculation(entries: dict[str, bytes]) -> None:
    workbook = ET.fromstring(entries["xl/workbook.xml"])
    calculation = workbook.find("x:calcPr", NS)
    if calculation is None:
        calculation = ET.SubElement(workbook, f"{{{MAIN_NS}}}calcPr")
    calculation.attrib.update(
        {"calcId": "0", "fullCalcOnLoad": "1", "forceFullCalc": "1"}
    )
    entries["xl/workbook.xml"] = ET.tostring(
        workbook, encoding="utf-8", xml_declaration=True
    )


def build_workbook(
    comparison_path: Path, output_path: Path, template_path: Path
) -> None:
    if not comparison_path.is_file():
        raise FileNotFoundError(
            f"Comparison data not found: {comparison_path}. "
            "Run collect_evidence.py --run all first."
        )
    if not template_path.is_file():
        raise FileNotFoundError(f"Workbook template not found: {template_path}")

    comparison = json.loads(comparison_path.read_text(encoding="utf-8"))
    with zipfile.ZipFile(template_path, "r") as source:
        infos = source.infolist()
        entries = {info.filename: source.read(info.filename) for info in infos}

    paths = _sheet_paths(entries)
    shared_strings = _shared_strings(entries)
    for sheet_name, group in SHEET_GROUPS.items():
        sheet = ET.fromstring(entries[paths[sheet_name]])
        _populate_evaluation_sheet(sheet, group, comparison, shared_strings)
        entries[paths[sheet_name]] = ET.tostring(
            sheet, encoding="utf-8", xml_declaration=True
        )

    recovery = ET.fromstring(entries[paths["PoC Recovery"]])
    _populate_recovery(recovery, comparison)
    entries[paths["PoC Recovery"]] = ET.tostring(
        recovery, encoding="utf-8", xml_declaration=True
    )
    _force_recalculation(entries)

    output_path.parent.mkdir(parents=True, exist_ok=True)
    temporary = tempfile.NamedTemporaryFile(
        prefix=f"{output_path.stem}.",
        suffix=".xlsx",
        dir=output_path.parent,
        delete=False,
    )
    temporary_path = Path(temporary.name)
    temporary.close()
    try:
        with zipfile.ZipFile(
            temporary_path, "w", compression=zipfile.ZIP_DEFLATED
        ) as destination:
            for info in infos:
                destination.writestr(info, entries[info.filename])
        os.replace(temporary_path, output_path)
    finally:
        if temporary_path.exists():
            temporary_path.unlink()


def main() -> int:
    script_directory = Path(__file__).resolve().parent
    package_root = script_directory.parent
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "comparison",
        nargs="?",
        type=Path,
        default=package_root / "results" / "comparison_data.json",
    )
    parser.add_argument(
        "output",
        nargs="?",
        type=Path,
        default=package_root / "results" / "UC01_comparison.xlsx",
    )
    parser.add_argument(
        "--template",
        type=Path,
        default=package_root / "templates" / "UC01_comparison_template.xlsx",
    )
    arguments = parser.parse_args()
    build_workbook(
        arguments.comparison.resolve(),
        arguments.output.resolve(),
        arguments.template.resolve(),
    )
    print(f"Wrote {arguments.output.resolve()}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: evaluation_layer/eval_methods/Solution wizard evaluation/use cases/UC01_audio_incident_evaluation/scripts/build_workbook.mjs
#!/usr/bin/env node
/**
 * Build the side-by-side UC01 review workbook with @oai/artifact-tool.
 *
 * Usage:
 *   node scripts/build_workbook.mjs
 *   node scripts/build_workbook.mjs results/comparison_data.json results/UC01_comparison.xlsx
 */

import fs from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { SpreadsheetFile, Workbook } from "@oai/artifact-tool";

const scriptDir = path.dirname(fileURLToPath(import.meta.url));
const packageRoot = path.resolve(scriptDir, "..");
const positionalArgs = process.argv.slice(2).filter((argument) => !argument.startsWith("--"));
const comparisonPath = path.resolve(
  positionalArgs[0] || path.join(packageRoot, "results", "comparison_data.json"),
);
const outputPath = path.resolve(
  positionalArgs[1] || path.join(packageRoot, "results", "UC01_comparison.xlsx"),
);
const templateMode = process.argv.includes("--template");
const previewArgument = process.argv.find((argument) => argument.startsWith("--preview-dir="));
const previewDir = previewArgument
  ? path.resolve(previewArgument.slice("--preview-dir=".length))
  : null;

const oracle = JSON.parse(
  await fs.readFile(path.join(packageRoot, "scenario_oracle.json"), "utf8"),
);
let comparison = { runs: {} };
try {
  comparison = JSON.parse(await fs.readFile(comparisonPath, "utf8"));
} catch {
  comparison = { runs: {} };
}

const workbook = Workbook.create();
workbook.worksheets.add("Instructions");
workbook.worksheets.add("Summary");
workbook.worksheets.add("EQ1 Requirements");
workbook.worksheets.add("EQ1 Diagnostics");
workbook.worksheets.add("EQ2 Configuration");
workbook.worksheets.add("EQ3 Package");
workbook.worksheets.add("EQ4 Execution");
workbook.worksheets.add("PoC Recovery");

const colors = {
  navy: "#17365D",
  blue: "#D9EAF7",
  paleBlue: "#EEF5FB",
  green: "#E2F0D9",
  red: "#FCE4D6",
  amber: "#FFF2CC",
  gray: "#E7E6E6",
  darkGray: "#595959",
  white: "#FFFFFF",
};

function titleBand(sheet, title, subtitle) {
  sheet.getRange("A1:N1").merge();
  sheet.getRange("A1").values = [[title]];
  sheet.getRange("A1:N1").format = {
    fill: colors.navy,
    font: { bold: true, color: colors.white, size: 16 },
    verticalAlignment: "center",
  };
  sheet.getRange("A1:N1").format.rowHeight = 28;
  sheet.getRange("A2:N2").merge();
  sheet.getRange("A2").values = [[subtitle]];
  sheet.getRange("A2:N2").format = {
    fill: colors.paleBlue,
    font: { italic: true, color: colors.darkGray, size: 10 },
    wrapText: true,
    verticalAlignment: "center",
  };
  sheet.getRange("A2:N2").format.rowHeight = 40;
  sheet.showGridLines = false;
}

function resultRows(group) {
  const run1 = comparison?.runs?.run_01?.results?.[group] || [];
  const run2 = comparison?.runs?.run_02?.results?.[group] || [];
  return {
    run1: new Map(run1.map((row) => [row.check_id, row])),
    run2: new Map(run2.map((row) => [row.check_id, row])),
  };
}

function writeEvaluationSheet(sheetName, group, checks, subtitle) {
  const sheet = workbook.worksheets.getItem(sheetName);
  titleBand(sheet, sheetName, subtitle);
  const headers = [
    "Scored?",
    "Check ID",
    "Oracle parameter",
    "Oracle expected value",
    "Oracle evidence/source",
    "Run 1 generated value",
    "Run 1 evidence",
    "Run 1 auto aid",
    "Run 1 verdict",
    "Run 2 generated value",
    "Run 2 evidence",
    "Run 2 auto aid",
    "Run 2 verdict",
    "Evaluator notes",
  ];
  sheet.getRange("A3:N3").values = [headers];
  sheet.getRange("A3:N3").format = {
    fill: colors.blue,
    font: { bold: true, color: colors.navy },
    wrapText: true,
    verticalAlignment: "center",
    borders: { preset: "all", style: "thin", color: "#B4C6E7" },
  };
  sheet.getRange("A3:N3").format.rowHeight = 34;

  const { run1, run2 } = resultRows(group);
  const rows = checks.map((check) => {
    const r1 = run1.get(check.id) || {};
    const r2 = run2.get(check.id) || {};
    const source = Array.isArray(check.source)
      ? check.source.join("; ")
      : Array.isArray(check.evidence_required)
        ? check.evidence_required.join("; ")
        : "";
    return [
      check.scored === false ? "No" : "Yes",
      check.id,
      check.parameter,
      typeof check.oracle_value === "string"
        ? check.oracle_value
        : JSON.stringify(check.oracle_value),
      source,
      r1.generated_value || "",
      r1.evidence || "",
      r1.automatic_result || "",
      r1.human_verdict || "",
      r2.generated_value || "",
      r2.evidence || "",
      r2.automatic_result || "",
      r2.human_verdict || "",
      [r1.evaluator_notes, r2.evaluator_notes].filter(Boolean).join(" | "),
    ];
  });
  const lastRow = 3 + rows.length;
  if (rows.length) {
    sheet.getRange(`A4:N${lastRow}`).values = rows;
    sheet.getRange(`A4:N${lastRow}`).format = {
      wrapText: true,
      verticalAlignment: "top",
      borders: { preset: "all", style: "thin", color: "#D9E1F2" },
    };
    sheet.getRange(`A4:N${lastRow}`).format.rowHeight = 72;
    sheet.getRange(`I4:I${lastRow}`).dataValidation = {
      rule: { type: "list", values: ["Yes", "No"] },
    };
    sheet.getRange(`M4:M${lastRow}`).dataValidation = {
      rule: { type: "list", values: ["Yes", "No"] },
    };
    for (const verdictRange of [`I4:I${lastRow}`, `M4:M${lastRow}`]) {
      sheet.getRange(verdictRange).conditionalFormats.add("cellIs", {
        operator: "equalTo",
        formula: '"Yes"',
        format: { fill: colors.green, font: { bold: true, color: "#375623" } },
      });
      sheet.getRange(verdictRange).conditionalFormats.add("cellIs", {
        operator: "equalTo",
        formula: '"No"',
        format: { fill: colors.red, font: { bold: true, color: "#9C0006" } },
      });
    }
    sheet.getRange(`H4:H${lastRow}`).format.fill = colors.gray;
    sheet.getRange(`L4:L${lastRow}`).format.fill = colors.gray;
  }
  const widths = [9, 13, 24, 34, 26, 34, 46, 13, 13, 34, 46, 13, 13, 28];
  widths.forEach((width, index) => {
    sheet.getRangeByIndexes(0, index, lastRow, 1).format.columnWidth = width;
  });
  sheet.freezePanes.freezeRows(3);
  sheet.freezePanes.freezeColumns(2);
  sheet.getRange(`A3:N${lastRow}`).format.autofitRows();
  return { firstRow: 4, lastRow };
}

const ranges = {};
ranges.eq1 = writeEvaluationSheet(
  "EQ1 Requirements",
  "EQ1",
  oracle.checks.EQ1,
  "Requirement capture recall. Review semantically equivalent values; an omitted oracle requirement remains a row and receives No.",
);
ranges.diag = writeEvaluationSheet(
  "EQ1 Diagnostics",
  "EQ1_diagnostics",
  oracle.checks.EQ1_diagnostics,
  "Unscored transparency diagnostic for unsupported commitments. Evidence is still mandatory.",
);
ranges.eq2 = writeEvaluationSheet(
  "EQ2 Configuration",
  "EQ2",
  oracle.checks.EQ2,
  "GAIK module/component constraints and behavior-changing options.",
);
ranges.eq3 = writeEvaluationSheet(
  "EQ3 Package",
  "EQ3",
  oracle.checks.EQ3,
  "A run is a valid solution package only when every mandatory EQ3 check receives Yes.",
);
ranges.eq4 = writeEvaluationSheet(
  "EQ4 Execution",
  "EQ4",
  oracle.checks.EQ4,
  "A run is a successful PoC only when every mandatory EQ4 check for the original unmodified PoC receives Yes.",
);

const recoverySheet = workbook.worksheets.getItem("PoC Recovery");
titleBand(
  recoverySheet,
  "PoC recovery diagnostic",
  "Unscored diagnostic. EQ4 remains based on the original unmodified PoC; recovery permits at most three wizard refinement attempts.",
);
recoverySheet.getRange("A3:H3").values = [[
  "Diagnostic",
  "Run 1 value",
  "Run 1 evidence",
  "Run 2 value",
  "Run 2 evidence",
  "Protocol value",
  "Scored?",
  "Interpretation",
]];
recoverySheet.getRange("A3:H3").format = {
  fill: colors.blue,
  font: { bold: true, color: colors.navy },
  wrapText: true,
  verticalAlignment: "center",
  borders: { preset: "all", style: "thin", color: "#B4C6E7" },
};
const recovery1 = comparison?.runs?.run_01?.recovery_diagnostic || {};
const recovery2 = comparison?.runs?.run_02?.recovery_diagnostic || {};
const displayValue = (value) => {
  if (value === true) return "Yes";
  if (value === false) return "No";
  if (value === null || value === undefined) return "";
  return String(value);
};
recoverySheet.getRange("A4:H7").values = [
  [
    "Original execution successful",
    displayValue(recovery1.initial_execution_successful),
    recovery1.evidence || "",
    displayValue(recovery2.initial_execution_successful),
    recovery2.evidence || "",
    "Yes or No",
    "No",
    "The original execution remains the EQ4 evidence source.",
  ],
  [
    "Refinement attempts to success",
    displayValue(recovery1.refinement_attempts_to_success),
    recovery1.evidence || "",
    displayValue(recovery2.refinement_attempts_to_success),
    recovery2.evidence || "",
    "0, 1, 2, 3, or N/A",
    "No",
    "0 means original success; N/A means never successful after three refinements.",
  ],
  [
    "Final execution successful",
    displayValue(recovery1.final_execution_successful),
    recovery1.evidence || "",
    displayValue(recovery2.final_execution_successful),
    recovery2.evidence || "",
    "Yes, No, or blank",
    "No",
    "Shows whether recovery ultimately produced an executable PoC.",
  ],
  [
    "Recovery status",
    displayValue(recovery1.status),
    recovery1.evidence || "",
    displayValue(recovery2.status),
    recovery2.evidence || "",
    "Protocol-controlled status",
    "No",
    "Stop after first success or after refinement attempt 3.",
  ],
];
recoverySheet.getRange("A4:H7").format = {
  wrapText: true,
  verticalAlignment: "top",
  borders: { preset: "all", style: "thin", color: "#D9E1F2" },
};
recoverySheet.getRange("A4:H7").format.rowHeight = 76;
[30, 18, 56, 18, 56, 25, 12, 48].forEach((width, index) => {
  recoverySheet.getRangeByIndexes(0, index, 7, 1).format.columnWidth = width;
});
recoverySheet.getRange("G4:G7").format.fill = colors.gray;
recoverySheet.freezePanes.freezeRows(3);
recoverySheet.showGridLines = false;

const instructions = workbook.worksheets.getItem("Instructions");
titleBand(
  instructions,
  "UC01 evaluation review workbook",
  "The workbook displays oracle and wizard evidence side by side. Automated aids never replace the evaluator's Yes/No verdict.",
);
instructions.getRange("A4:B19").values = [
  ["Step", "Action"],
  ["1", "For each run, select its generated_package folder; then send initial_prompt.txt and use scripted_answers.md. No path or provenance metadata needs to be entered. At every routine confirmation, answer Yes and proceed without changes."],
  ["2", "Save each conversation and complete generated package under runs/run_01 and runs/run_02."],
  ["3", "Run run_poc_evaluation.py with attempt 0. This original, unmodified execution is the only execution used for EQ4."],
  ["4", "If attempt 0 fails, return its execution evidence to the wizard and run refinement attempts 1-3 in order. Stop at first success."],
  ["5", "If all three refinements fail, poc_recovery.json records N/A. Refined success never overwrites EQ4."],
  ["6", "Run collect_evidence.py --run all. Use --use-llm only to locate/align evidence; the LLM does not decide verdicts."],
  ["7", "Rebuild this workbook with build_workbook.mjs so the Run 1 and Run 2 value/evidence columns are populated."],
  ["8", "For every scored row, compare the oracle value with the generated value and inspect the adjacent evidence."],
  ["9", "Enter Yes only when the generated result satisfies the accepted meaning. Enter No when it does not or when it is missing."],
  ["10", "For missing requirements, keep NOT FOUND and its search evidence; do not remove the row. It remains in the denominator."],
  ["11", "The Auto aid column reports deterministic pass/fail/needs_review evidence. It is not the final verdict."],
  ["12", "Run score_workbook.mjs after every scored row in both runs has a Yes or No verdict."],
  ["Evidence rule", "Every verdict must be supported by an adjacent JSON path/value, conversation line, file check, command/exit code, log excerpt, or output comparison."],
  ["EQ1", "Requirement capture recall = pooled Yes verdicts / pooled scored requirement checks across two runs."],
  ["EQ2", "Configuration constraint satisfaction = pooled Yes verdicts / pooled configuration checks across two runs."],
];
instructions.getRange("A4:B19").format = {
  wrapText: true,
  verticalAlignment: "top",
  borders: { preset: "all", style: "thin", color: "#D9E1F2" },
};
instructions.getRange("A4:B4").format = {
  fill: colors.blue,
  font: { bold: true, color: colors.navy },
};
instructions.getRange("A4:A19").format.columnWidth = 18;
instructions.getRange("B4:B19").format.columnWidth = 110;
instructions.getRange("A4:B19").format.autofitRows();
instructions.freezePanes.freezeRows(3);

const summary = workbook.worksheets.getItem("Summary");
titleBand(
  summary,
  "UC01 evaluation summary",
  "Scores appear only after all required human verdicts are entered. EQ3 and EQ4 use an all-mandatory-checks pass rule per run.",
);
summary.getRange("A4:H4").values = [[
  "EQ",
  "Metric",
  "Run 1",
  "Run 2",
  "Combined score",
  "Completed verdicts",
  "Required verdicts",
  "Interpretation",
]];
summary.getRange("A5:H8").values = [
  ["EQ1", "Requirement capture recall", "", "", "", "", "", "Pooled row-level recall"],
  ["EQ2", "Configuration constraint satisfaction", "", "", "", "", "", "Pooled constraint satisfaction"],
  ["EQ3", "Valid solution package rate", "", "", "", "", "", "Run passes only if all EQ3 rows are Yes"],
  ["EQ4", "PoC execution success rate", "", "", "", "", "", "Run passes only if all EQ4 rows are Yes"],
];
summary.getRange("A4:H8").format = {
  wrapText: true,
  verticalAlignment: "center",
  borders: { preset: "all", style: "thin", color: "#B4C6E7" },
};
summary.getRange("A4:H4").format = {
  fill: colors.blue,
  font: { bold: true, color: colors.navy },
};

function eqRowFormula(summaryRow, sheetName, range, binary) {
  const r1 = `'${sheetName}'!I${range.firstRow}:I${range.lastRow}`;
  const r2 = `'${sheetName}'!M${range.firstRow}:M${range.lastRow}`;
  const count = range.lastRow - range.firstRow + 1;
  const completed1 = `COUNTIF(${r1},"Yes")+COUNTIF(${r1},"No")`;
  const completed2 = `COUNTIF(${r2},"Yes")+COUNTIF(${r2},"No")`;
  if (binary) {
    summary.getRange(`C${summaryRow}`).formulas = [[`=IF(${completed1}=${count},IF(COUNTIF(${r1},"No")=0,1,0),"")`]];
    summary.getRange(`D${summaryRow}`).formulas = [[`=IF(${completed2}=${count},IF(COUNTIF(${r2},"No")=0,1,0),"")`]];
    summary.getRange(`E${summaryRow}`).formulas = [[`=IF(AND(ISNUMBER(C${summaryRow}),ISNUMBER(D${summaryRow})),AVERAGE(C${summaryRow}:D${summaryRow}),"")`]];
  } else {
    summary.getRange(`C${summaryRow}`).formulas = [[`=IF(${completed1}=${count},COUNTIF(${r1},"Yes")/${count},"")`]];
    summary.getRange(`D${summaryRow}`).formulas = [[`=IF(${completed2}=${count},COUNTIF(${r2},"Yes")/${count},"")`]];
    summary.getRange(`E${summaryRow}`).formulas = [[`=IF(AND(ISNUMBER(C${summaryRow}),ISNUMBER(D${summaryRow})),(COUNTIF(${r1},"Yes")+COUNTIF(${r2},"Yes"))/${count * 2},"")`]];
  }
  summary.getRange(`F${summaryRow}`).formulas = [[`=${completed1}+${completed2}`]];
  summary.getRange(`G${summaryRow}`).values = [[count * 2]];
}

eqRowFormula(5, "EQ1 Requirements", ranges.eq1, false);
eqRowFormula(6, "EQ2 Configuration", ranges.eq2, false);
eqRowFormula(7, "EQ3 Package", ranges.eq3, true);
eqRowFormula(8, "EQ4 Execution", ranges.eq4, true);
summary.getRange("C5:E8").format.numberFormat = [["0%"]];
summary.getRange("C5:E8").format.fill = colors.paleBlue;
summary.getRange("A10:H14").values = [
  ["Diagnostic", "Run 1", "Run 2", "Reporting", "", "", "", "Not included in headline scores"],
  ["Unsupported assumptions", "", "", "Yes/No diagnostic verdict counts", "", "", "", "Review provider/model, retention, SLA/scale commitments"],
  ["Refinement attempts to success", displayValue(recovery1.refinement_attempts_to_success), displayValue(recovery2.refinement_attempts_to_success), "0, 1, 2, 3, or N/A", "", "", "", "Original EQ4 is unchanged"],
  ["Recovery status", displayValue(recovery1.status), displayValue(recovery2.status), "Initially successful, recovered, pending, or unsuccessful", "", "", "", "Maximum three refinement attempts"],
  ["Note", "", "", "", "", "", "", "Blank verdicts mean the evaluation is incomplete, not zero performance."],
];
summary.getRange("B11").formulas = [[`="Yes: "&COUNTIF('EQ1 Diagnostics'!I${ranges.diag.firstRow}:I${ranges.diag.lastRow},"Yes")&"; No: "&COUNTIF('EQ1 Diagnostics'!I${ranges.diag.firstRow}:I${ranges.diag.lastRow},"No")`]];
summary.getRange("C11").formulas = [[`="Yes: "&COUNTIF('EQ1 Diagnostics'!M${ranges.diag.firstRow}:M${ranges.diag.lastRow},"Yes")&"; No: "&COUNTIF('EQ1 Diagnostics'!M${ranges.diag.firstRow}:M${ranges.diag.lastRow},"No")`]];
summary.getRange("A10:H14").format = {
  wrapText: true,
  borders: { preset: "all", style: "thin", color: "#D9E1F2" },
};
summary.getRange("A10:H10").format = {
  fill: colors.amber,
  font: { bold: true, color: colors.navy },
};
[10, 28, 14, 14, 17, 17, 17, 52].forEach((width, index) => {
  summary.getRangeByIndexes(0, index, 14, 1).format.columnWidth = width;
});
summary.getRange("A4:H14").format.autofitRows();
summary.freezePanes.freezeRows(4);

const summaryInspection = await workbook.inspect({
  kind: "table",
  range: "Summary!A1:H14",
  include: "values,formulas",
  tableMaxRows: 14,
  tableMaxCols: 8,
});
console.log(summaryInspection.ndjson);

const formulaErrors = await workbook.inspect({
  kind: "match",
  searchTerm: "#REF!|#DIV/0!|#VALUE!|#NAME\\?|#N/A",
  options: { useRegex: true, maxResults: 300 },
  summary: "final formula error scan",
});
if (formulaErrors.ndjson && !formulaErrors.ndjson.includes('"matches":[]')) {
  console.log(formulaErrors.ndjson);
}

if (previewDir) {
  await fs.mkdir(previewDir, { recursive: true });
  for (const sheetName of [
    "Instructions",
    "Summary",
    "EQ1 Requirements",
    "EQ1 Diagnostics",
    "EQ2 Configuration",
    "EQ3 Package",
    "EQ4 Execution",
    "PoC Recovery",
  ]) {
    const preview = await workbook.render({
      sheetName,
      autoCrop: "all",
      scale: 1,
      format: "png",
    });
    const safeName = sheetName.toLowerCase().replaceAll(" ", "_");
    await fs.writeFile(
      path.join(previewDir, `${safeName}.png`),
      new Uint8Array(await preview.arrayBuffer()),
    );
  }
  console.log(`Rendered workbook sheets to ${previewDir}`);
}

await fs.mkdir(path.dirname(outputPath), { recursive: true });
const output = await SpreadsheetFile.exportXlsx(workbook);
await output.save(outputPath);
console.log(`Wrote ${outputPath}${templateMode ? " (template mode)" : ""}`);
